### ISSUE TITLE
Small issue with lyd_new_path() + ly_errmsg()

### ISSUE THREAD

**Issue by rwestphal** [CESNET/libyang#520]
Hi,

When using `lyd_new_path()` to create an optional leaf with an invalid value, the string returned by `ly_errmsg()` depends on whether the optional leaf already exists in the data tree or not.

As an example, please take a look at this optional leaf of my YANG module:
```
          leaf password {
            type string {
              length "1..16";
            }
            mandatory true;
            description
              "Authentication string.";
          }
```
(Full YANG module: https://gist.github.com/rwestphal/2fc93f3f928fee1127c2913977c00e8d)

Note the length constraint: `"1..16"`.

Now using the CLI of my program I got this:
```
ripd> enable                                                                                                                                                                                       
ripd# conf t    
ripd(config)# interface eth0                     
ripd(config-if)# show configuration candidate xml with-defaults
<lib xmlns="http://frrouting.org/yang/interface">
  <interface>
    <name>eth0</name>
    <vrf>Default-IP-Routing-Table</vrf>
    <rip xmlns="http://frrouting.org/yang/ripd">                           
      <split-horizon>simple</split-horizon>
      <v2-broadcast>false</v2-broadcast>
      <version-receive>unspecified</version-receive>                                                      
      <version-send>unspecified</version-send>
      <authentication>
        <type>none</type>
      </authentication>
    </rip>
  </interface>
</lib>
ripd(config-if)#
ripd(config-if)# ip rip authentication string XXXXXXXXXXXXXXXXXXXXXXXXXXXXX
% failed to edit candidate configuration

Failed to create node "password" as a child of "authentication".
ripd(config-if)#
ripd(config-if)# ip rip authentication string XXXXXXXXXXX                  
ripd(config-if)#
ripd(config-if)# show configuration candidate xml with-defaults
<lib xmlns="http://frrouting.org/yang/interface">
  <interface>
    <name>eth0</name>
    <vrf>Default-IP-Routing-Table</vrf>
    <rip xmlns="http://frrouting.org/yang/ripd">
      <split-horizon>simple</split-horizon>
      <v2-broadcast>false</v2-broadcast>
      <version-receive>unspecified</version-receive>
      <version-send>unspecified</version-send>
      <authentication>
        <type>none</type>
        <password>XXXXXXXXXXX</password>
      </authentication>
    </rip>
  </interface>
</lib>
ripd(config-if)#
ripd(config-if)# ip rip authentication string XXXXXXXXXXXXXXXXXXXXXXXXXXXXX
% failed to edit candidate configuration

Value "XXXXXXXXXXXXXXXXXXXXXXXXXXXXX" does not satisfy the constraint "1..16" (range, length, or pattern).
```

In short, if the `password` leaf doesn't exist and we try to create it using an invalid value, the error returned by `lyd_new_path()` is this:
```
Failed to create node "password" as a child of "authentication".
```

However, if the `password` leaf does already exist and we try to update its value using an invalid string, the error returned by `lyd_new_path()` is this:
```
Value "XXXXXXXXXXXXXXXXXXXXXXXXXXXXX" does not satisfy the constraint "1..16" (range, length, or pattern).
```

I think the first error message is too generic and not very helpful, which leads me to ask: would it be possible to return the second error message regardless if the `password` leaf exists or not?

Regards,
Renato.

**Comment by michalvasko** [comment]
Hi Renato,
this is natural due to the fact that node must first be created and then the value is checked (I received the pattern error when the node did not exist and vice versa so I am going to assume this is what happens to you and you just described it wrong). I got your desired behaviour when using `LYD_OPT_PATH_UPDATE` flag for calling `lyd_new_path` when the node already existed so maybe you want to always use it?

Regards,
Michal

**Comment by rwestphal** [comment]
@michalvasko I'm always using `lyd_new_path()` with the `LYD_OPT_PATH_UPDATE` flag.

I wrote this simple program to better illustrate my problem: https://gist.github.com/rwestphal/36d26cc7b37e151c38ceccc7d374e1c4

Here's the output (the valid range for `default-metric` is `1..16`):
```
$ gcc -Wall -g3 -O0 lyd_new_path_errmsg.c -o lyd_new_path_errmsg -lyang
$ ./lyd_new_path_errmsg 
Setting '/frr-ripd:ripd/instance/default-metric' to '18':
Error: lyd_new_path(): Failed to create node "default-metric" as a child of "instance".

Setting '/frr-ripd:ripd/instance/default-metric' to '2':
Ok

Setting '/frr-ripd:ripd/instance/default-metric' to '18':
Error: lyd_new_path(): Value "18" does not satisfy the constraint "1..16" (range, length, or pattern).
```

I'm receiving the pattern error only when the leaf already exists in the data node (not when it doesn't exist). I wouldn't classify this as a bug but it would be nice to have the second error message in both cases for consistency.

Regards,
Renato.

**Comment by michalvasko** [comment]
Hi Renato,
please carefully read [this doc section](https://netopeer.liberouter.org/doc/libyang/devel/howtologger.html) and try this modification of your program.

```c
#include <stdlib.h>
#include <stdio.h>
#include <stdint.h>
#include <string.h>
#include <assert.h>
#include <err.h>

#include <libyang/libyang.h>

static struct ly_ctx *ly_ctx;

static void validate(struct lyd_node **config)
{
	if (lyd_validate(config, LYD_OPT_STRICT | LYD_OPT_CONFIG, ly_ctx) != 0)
		errx(1, "lyd_validate(): %s", ly_errmsg(ly_ctx));
}

static void ly_log_cb(LY_LOG_LEVEL level, const char *msg, const char *path)
{
	/* do nothing */
}

static void
print_all_errors(struct ly_ctx *ctx)
{
    struct ly_err_item *ei;

    for (ei = ly_err_first(ctx); ei; ei = ei->next) {
        printf("Error: %s\n", ei->msg);
    }

    ly_err_clean(ctx, NULL);
}

static void update(struct lyd_node *config, const char *path, const char *value)
{
	struct lyd_node *node;

	printf("Setting '%s' to '%s':\n", path, value);

	ly_errno = 0;
	node = lyd_new_path(config, ly_ctx, path, (void *)value, 0,
			    LYD_PATH_OPT_UPDATE);
	if (node == NULL && ly_errno) {
        print_all_errors(ly_ctx);
		printf("Error: lyd_new_path() failed\n");
    } else
		printf("Ok\n");
	printf("\n");
}

int main(int argc, char **argv)
{
	const struct lys_module *module;
	struct lyd_node *config = NULL;

	/* initialization */
	ly_ctx = ly_ctx_new(NULL, 0);
	if (!ly_ctx)
		errx(1, "ly_ctx_new(): %s", ly_errmsg(ly_ctx));
    ly_log_options(LY_LOLOG | LY_LOSTORE);
	ly_set_log_clb(ly_log_cb, 0);
	module = lys_parse_path(ly_ctx, "frr-ripd.yang", LYS_IN_YANG);
	if (module == NULL) {
        print_all_errors(ly_ctx);
		errx(1, "lys_parse_path() failed");
    }
	validate(&config);

	/* perform test */
	update(config, "/frr-ripd:ripd/instance/default-metric", "18");
	update(config, "/frr-ripd:ripd/instance/default-metric", "2");
	update(config, "/frr-ripd:ripd/instance/default-metric", "18");

	/* cleanup */
	lyd_free_withsiblings(config);
	ly_ctx_remove_module(module, NULL);
	ly_ctx_destroy(ly_ctx, NULL);

	return 0;
}
```

Just a note, I think it is always simplest to use the logging callback (either the default or a custom one, but not empty like yours!). However, there are other options to suit hopefully all use-cases I can think of (like I demonstrated in the app).

Regards,
Michal

**Comment by rwestphal** [comment]
Thanks a lot Michal. It never occurred to me that a single library call could return multiple errors, but it makes perfect sense. I'll consider using the logging callback in my program or resort to `ly_err_first()` and friends if I need more flexibility. Thanks again!
